pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/FunctionsBase.java

```diff
@@ -276,6 +276,26 @@ public FunctionState getFunctionState(final @PathParam("tenant") String tenant,
         return functions.getFunctionState(tenant, namespace, functionName, key, clientAppId(), clientAuthData());
     }
 
+    @POST
+    @ApiOperation(
+            value = "Put the state associated with a Pulsar Function"
+    )
+    @ApiResponses(value = {
+            @ApiResponse(code = 400, message = "Invalid request"),
+            @ApiResponse(code = 403, message = "The requester doesn't have admin permissions"),
+            @ApiResponse(code = 404, message = "The function does not exist"),
+            @ApiResponse(code = 500, message = "Internal server error")
+    })
+    @Path("/{tenant}/{namespace}/{functionName}/state/{key}")
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public void putFunctionState(final @PathParam("tenant") String tenant,
+                                 final @PathParam("namespace") String namespace,
+                                 final @PathParam("functionName") String functionName,
+                                 final @PathParam("key") String key,
+                                 final @FormDataParam("state") String stateJson) {
```

Comment:
I don't think we should use "FormData", that is usually used for submitting via forms on a website.  You just have "FunctionState state" as input here

Reply:
I know we use "FormData" everything where else but I don't think that is correct

Reply:
What will happen if we include more items in the future? How will it know which keys map to which data?

Reply:
In that case, the input should be a json object that contains all the fields.  That is usually what people are doing nowadays.  For example:

https://www.vaultproject.io/api/secret/databases/cassandra.html#sample-payload

Reply:
Take the example of 'big objects'. In future versions we might introduce a way to stream the big object via an inputstream. Thus a single object might not work for this case?